### PR TITLE
Add typed events for all user hooks and legacy events

### PR DESCRIPTION
**Pull request by ChristophWurst**
There is a lot of dark magic we use for user change propagation. That is why removing the emit stuff in https://github.com/nextcloud/server/pull/18225 failed on CI terribly. This needs a more in-depth analysis.

To still get the new events in for 18 I created a modified version where I leave the old mechanisms in place and just convert the old stuff to the new (and not the other way around). The cleanup will follow for 19.

Supersedes https://github.com/nextcloud/server/pull/18225.

### CODE REVIEW

File: lib/public/User/Events/BeforeUserLoggedInWithCookieEvent.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	 */
+	public function __construct(string $username) {
+		parent::__construct();
+		$this->username = $username;
+	}
+
+	/**
+	 * @since 18.0.0
+	 */
+	public function getUsername(): IUser {
```
Comment (georgehrke) [inline lib/public/User/Events/BeforeUserLoggedInWithCookieEvent.php]: Constructor says `string`

File: lib/public/User/Events/BeforeUserLoggedInEvent.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	public function __construct(string $username, string $password) {
+		parent::__construct();
+		$this->username = $username;
+		$this->password = $password;
+	}
+
+	/**
+	 * @since 18.0.0
+	 */
+	public function getUsername(): IUser {
```
Comment (georgehrke) [inline lib/public/User/Events/BeforeUserLoggedInEvent.php]: Constructor says `string`

**Review by skjnldsv — APPROVED** [review]
Crazy work!
We shall document this!

**Comment by ChristophWurst** [comment]
> We shall document this!

on my todo :)

**Review by rullzer — APPROVED** [review]
Lets do this!

**Comment by MorrisJobke** [comment]
ref #14552

**Pull request merged**